### ISSUE TITLE
Subexpression type narrowing in match statement doesn't narrow as expected

### ISSUE THREAD

**Issue by sirrus233** [microsoft/pyright#4995]
I'm trying out pyright `1.1.304`, specifically very excited for #4971 , and I noticed that the type narrowing is not working quite as I expected. Consider:

```
def foo(x: int | None, y: int | None):
    match (x, y):
        case (None, None):
            ...
        case (x, None):  # x is not narrowed, still type int | None (???)
            ...
        case (None, y):  # y is narrowed to int, as expected
            ...
        case (x, y):     # x and y are both narrowed to int, as expected
            ...
 ```

I would expect `x` in the second case to be narrowed to `int`, since if it were `None` we should already be caught by the first case. Am I missing something here, or is this a bug?

I see the same behavior with CLI pyright, and with the preview version of the VS Code extension.

**Comment by erictraut** [comment]
Pyright is working as designed here. The first pattern is matched only in the case where _both_ `x` and `y` are `None`. If the first pattern is _not_ matched, it doesn't mean that `x` is not `None.

This is the same as an `if` statement that includes and `and`:

```python
def foo(x: int | None, y: int | None):
    if x is None and y is None:
        reveal_type(x)  # None
        reveal_type(y)  # None
    else:
        reveal_type(x)  # int | None
        reveal_type(y)  # int | None
```

I recently added a feature that allows a match subject expressions to be narrowed in the case where the subject expression is a tuple and the pattern narrows precisely one of the entries in the tuple. For example, if you use the pattern `(x, None)`, that narrows the second entry to `None` but does not narrow the first entry. I decided not to support the case where a pattern narrows multiple entries because that results in a combinatoric explosion. It produces a narrowed type consisting of a union of n! tuples, where n is the number of narrowed entries. 

In your example, the initial subject type is `tuple[int | None, int | None]`. The negative (fall-through) type after the first case statement could theoretically be narrowed to `tuple[int, int | None] | tuple[int | None, None]`. You can see how this gets unwieldy very quickly. With three narrowed entries, it would result in 6 entries in the union, etc.

I therefore decided to draw the line at one narrowed entry, which is a common use case. That's why you're seeing the narrowing behavior for `(x, None)` and `(None, y)`. Prior to pyright 1.1.304, you would have seen no narrowing in these cases. Also note that mypy and other type checkers don't do anything like this.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.374.
